qiskit/transpiler/passes/optimization/optimize_annotated.py

```diff
@@ -0,0 +1,188 @@
+                while isinstance(cur, AnnotatedOperation):
+                    modifiers.extend(cur.modifiers)
+                    cur = cur.base_op
+                canonical_modifiers = _canonicalize_modifiers(modifiers)
+                if len(canonical_modifiers) > 0:
+                    # this is still an annotated operation
+                    node.op.base_op = cur
+                    node.op.modifiers = canonical_modifiers
+                else:
+                    # no need for annotated operations
+                    node.op = cur
+                did_something = True
+        return dag, did_something
+
+    def _recursively_process_definitions(self, op: Operation) -> bool:
+        """
+        Recursively applies optimizations to op's definition (or to op.base_op's
+        definition if op is an annotated operation).
+        Returns True if did something.
+        """
+
+        # If op is an annotated operation, we descend into its base_op
+        if isinstance(op, AnnotatedOperation):
+            return self._recursively_process_definitions(op.base_op)
+
+        # Similar to HighLevelSynthesis transpiler pass, we do not recurse into a gate's
+        # `definition` for a gate that is supported by the target or in equivalence library.
+
+        controlled_gate_open_ctrl = isinstance(op, ControlledGate) and op._open_ctrl
+        if not controlled_gate_open_ctrl:
+            inst_supported = (
+                self._target.instruction_supported(operation_name=op.name)
+                if self._target is not None
+                else op.name in self._device_insts
+            )
+            if inst_supported or (self._equiv_lib is not None and self._equiv_lib.has_entry(op)):
+                return False
+
+        try:
+            # extract definition
+            definition = op.definition
+        except TypeError as err:
+            raise TranspilerError(
+                f"OptimizeAnnotated was unable to extract definition for {op.name}: {err}"
+            ) from err
+        except AttributeError:
+            # definition is None
+            definition = None
+
+        if definition is None:
+            raise TranspilerError(f"OptimizeAnnotated was unable to optimize {op}.")
+
+        definition_dag = circuit_to_dag(definition, copy_operations=False)
+        definition_dag, opt = self._run_inner(definition_dag)
+
+        if opt:
+            # We only update a gate's definition if it was actually changed.
+            # This is important to preserve non-annotated singleton gates.
+            op.definition = dag_to_circuit(definition_dag)
+
+        return opt
+
+    def _recurse(self, dag) -> Tuple[DAGCircuit, bool]:
+        """
+        Recursively handles gate definitions.
+        Returns True if did something.
+        """
+
+        did_something = False
+
+        # Don't do anything else if processing only top-level
+        if self._top_level_only:
+            return dag, did_something
```

Comment:
Can this ever be True? Aren't we guarding against this at L89? I would just remove this to simplify the logic a bit.

Reply:
Removed.